src/missions/mission_framework.rs

```diff
@@ -0,0 +1,51 @@
+trait State {
+    fn on_enter(&self);
+    fn on_periodic(&self);
+    fn on_exit(&self);
+    fn next_state(&self) -> Option<Box<dyn State>>; // return trait object
+}
+
+
+struct EmptyState;
+
+impl State for EmptyState {
+    fn on_enter(&self) {
+        stub!(); 
+    }
+    fn on_periodic(&self) {
+        stub!();
+    }
+    fn on_exit(&self) {
+        stub!();
+    }
+    fn next_state(&self) -> Option<Box<dyn State>> {
+        None
+    }
+
+}
+
+
+trait Mission {
+    type InitialState: State;
+
+    fn new(initial_state: Self::InitialState) -> Self;
+    fn current_state(&self) -> Option<&Box<dyn State>>;
+    fn set_current_state(&mut self, state: Box<dyn State>);
+
+    // Default implementation
+    fn run(&mut self) {
```

Comment:
I'm liking the use of generics and traits, but I'm not sure this is our cleanest approach to this. We probably shouldn't be using functions to pass up values from the implementer. I worked on the problem a bit and I'm thinking maybe we just want scoped functions instead of traits and structs? e.g.
```rust
mod basic_mission {
    use std::error::Error;

    use crate::State;

    pub async fn run(state: Box<dyn State>) -> Result<Option<Box<dyn State>>, Box<dyn Error>> {
        unimplemented!()
    }

    pub async fn run_all(state: Box<dyn State>) -> Result<(), Box<dyn Error>> {
        let mut state = Option::from(state);
        while let Some(cur_state) = state {
            state = run(cur_state).await?;
        }
        Ok(())
    }
}
``` 
And just provide a couple variants (atm all that comes to mind is a debug log version).